pkg/importer/spanner/sysl.arrai

```diff
@@ -0,0 +1,128 @@
+    ((entity.foreign_keys => ((.foreign_keys where .attribute = attr.name)
+        => .reference_attribute orderby .)) where .);
+
+# hasEntityPattern checks whether an entity has associated patterns.
+let hasEntityPattern = \entity \model
+    entity.primary_key | entity.foreign_keys | entity.cluster | model.indexes;
+
+# sortingOrder determines and appends sorting order.
+let sortingOrder = \e \attr
+    let re = //re.compile($`${attr.name}\((asc|desc)\)`);
+    let keyOrder = e.primary_key => (re.match(.)(0)?(1)?:{} rank (:.@)) where .;
+    cond keyOrder {{x}: $`, ~${x}`};
+
+# compareColumnOrder compares the col order between primary key and table.
+let compareColumnOrder = \entity
+    let pk = (entity.primary_key orderby .) >> //seq.split('(', .) >> .(0);
+    pk !(<=) (entity.attributes >> .name);
+
+# attributePatterns generates the patterns for an attribute.
+let attributePatterns = \entity \attr
+    let options = cond {
+        attr.options:
+            let opt = //seq.split('=',attr.options);
+            $`${opt(0)}='${opt(1)}'`,
+        };
+    let pk = cond {
+        (entity.primary_key where //seq.contains(attr.name, .)):
+            '~pk' ++ sortingOrder(entity, attr)
+    };
+    let fk = cond {
+        (entity.foreign_keys => ((.foreign_keys where .attribute = attr.name) =>
+        .reference_attribute orderby .)) where .:
+            '~fk'
+    };
+    let length = cond {
+        attr.length = 'MAX': '~max',
+    };
+    let hexPrefix = cond {
+        //seq.has_prefix('0x', attr.length): '~hex',
+    };
+    //seq.join(', ', ([options, pk, fk, length, hexPrefix] where .@item != ''));
+
+# entityPatterns generates the patterns for an entity.
+let entityPatterns = \entity \model
+    let pk = cond {
+        entity.primary_key count > 1 && compareColumnOrder(entity):
+            $`primary_key="${entity.primary_key orderby .::, }"`,
+    };
+    let cluster = cond {
+        entity.cluster:
+            //seq.join('', entity.cluster >>
+            $`interleave_in_parent="${.interleaved_in}", interleave_on_delete="${.on_delete}"`),
+    };
+    let fk = cond {
+        entity.foreign_keys:
+            $`foreign_keys=[${//seq.join(', ',(entity.foreign_keys =>
+                \keys $`["constraint:${keys.constraint_name}","columns:${//seq.join(', ', (keys.foreign_keys => .attribute) orderby .)}"]`) orderby .)}]`,
+    };
+    let indices = cond {
+        (model.indexes where .table_name = entity.name):
+            $`indexes=[${//seq.join(', ', ((model.indexes orderby .) where .@item.table_name = entity.name) >>
+                $`"name:${.table_name}","unique:${.unique}","null_filtered:${.nullfiltered}","key_parts:${//seq.join(', ', .key_part orderby .)}${cond {.storing_col: $`","storing:${//seq.join(', ',.storing_col)}`}}${cond {.interleaved_table: $`","interleave_in:${.interleaved_table}`}}`)}"]`,
+    };
+    //seq.join(', ', ([pk, cluster, fk, indices] where .@item != ''));
```

Comment:
Nit: again, I think it would be cleaner overall to have this function just produce an array of pattern strings, and leaves the joining to whatever function adds the `[...]` brackets around the result (i.e. `appendEntityPatterns`).
```suggestion
    [pk, cluster, fk, indices] where .@item != '';
```